rfc/rfc-48/rfc-48.md

```diff
@@ -0,0 +1,174 @@
+
+The purpose of LogCompaction is to reduce the write amplification caused due to compaction.
+
+
+## Background
+
+To stitch the log blocks together and to write a new compacted log block, clustering can also be used, 
+but it would not make sense to add replacecommit which actually works at a base file to merge the blocks 
+and complicate the existing logic. Instead existing compaction logic can be enhanced by adding LogCompaction action. 
+The schedule and execution of LogCompaction will be similar to compaction but it would create log blocks or files after merging them.
+
+
+## Implementation
+
+LogCompaction action needs to be added to the HoodieTimeline class. 
+LogCompaction is a table service and similar to other table services, 
+it is scheduled and can be executed either inline or async.
+
+
+Consider the following example: Let there be log blocks from 1 till 3. Now, 
+if LogCompaction is scheduled to run for these log blocks it creates a 
+<COMMIT_TIME>.logcompaction.requested on the timeline.
+
+
+Now LogCompaction executes and merges all the 3 log blocks and creates a new compacted 
+log block .log.4. When reading the log blocks, AbstractHoodieLogRecordReader sees 
+following 4 log blocks, but it will consider only log block 4.
+
+![](base_case.jpeg)
+
+
+Compacted log blocks are special type of HoodieLogBlock and they can be differentiated 
+based on the header key COMPACTED_BLOCKS_INSTANTS, which basically contains the list of
+log block's instants this compacted blocks contain.
+
+
+Along with the creation of a new command block type, rollback logic also need to be changed
+to handle multiwriter scenarios. Consider an example where a writer wrote a log block and could 
+not place a completed commit. Later another writer starts and appends to the previous invalid 
+block and succeeds. Now if we issue a rollback for the original commit that was killed, 
+the rollback block is not exactly on top of the invalid block. 
+To handle this Rollback logic needs to consider all the log blocks.
+
+
+Depending on the execution, any of the following scenarios can occur.
+#### Scenario 1: 
+
+Out of order merged blocks
+LogCompaction is scheduled and by the time it appended a block, another block is added by another writer.
+This can happen if logcompaction is running in async.
+
+![](async_logcompaction_issues.jpeg)
+
+#### Scenario 2:
+
+LogCompaction operation is failed, so a rollback is issued for this block. Here the rollback can be next block or can
+comes at a later point of time.
+
+![](rollback_block_scenario.jpeg)
+
+#### Scenario 3:
+
+Merged block itself is larger than the default block size of 256MB. 
+Single LogCompaction operation can create multiple log blocks.
+
+![](huge_data_after_logcompaction_written_to_multiple_blocks.jpeg)
+
+#### Scenario 4:
+
+LogCompaction is executing and before committing compaction starts running on the same file group.
```

Comment:
why do we even let log compaction proceed if there is a compaction scheduled?
or alternatively, should we think about not scheduling any compaction if log compaction is inflight for a given file slice. 
similarly, do not schedule any log compaction is compaction is scheduled for a given file slice?